File: SourceCode/AspCoreVersion/src/ShopAware.Core/DataObjects/DrugEventItem.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace ShopAware.Core
{
    [Obsolete]
    public class DrugEventItem
    {
        #region Member Variables

        public Dictionary<string, object> AllData = new Dictionary<string, object>();

        #endregion

        #region Properties

        public string SafetyReportID { get; set; }

        public string ReceivedDateString { get; set; }

        public string SenderOrganization { get; set; }

        public string PatientSex { get; set; }

        public string PatientAge { get; set; }

        public string PatientWeight { get; set; }

        public List<string> PatientReactions { get; set; }

        public string DrugStartDate { get; set; }

        public string DrugEndDate { get; set; }

        public string Doseage { get; set; }

        public List<string> Manufactures { get; set; }

        public List<string> BrandNames { get; set; }

        public List<string> GenericNames { get; set; }

        #endregion

        #region Constructors

        public DrugEventItem()
        {
            // VBConversions Note: Non-static class variable initialization is below.  Class variables cannot be initially assigned non-static values in C#.
            PatientReactions = new List<string>();
            Manufactures = new List<string>();
            BrandNames = new List<string>();
            GenericNames = new List<string>();
        }

        private DrugEventItem(Dictionary<string, object> jsonItem)
        {
            // VBConversions Note: Non-static class variable initialization is below.  Class variables cannot be initially assigned non-static values in C#.
            PatientReactions = new List<string>();
            Manufactures = new List<string>();
            BrandNames = new List<string>();
            GenericNames = new List<string>();

            Fetch(jsonItem);
        }

        #endregion

        #region Public Methods

        public static DrugEventItem Fill(Dictionary<string, object> item)
        {
            return new DrugEventItem(item);
        }

        #endregion

        #region Private Methods

        private void CheckKeys(string key, object value)
        {
            switch (key)
            {
                case "safetyreportid":
                    SafetyReportID = value.ToString();
                    break;
                case "receivedate":
                    ReceivedDateString = value.ToString();
                    break;
                case "senderorganization":
                    SenderOrganization = value.ToString();
                    break;
                case "patientonsetage":
                    PatientAge = value.ToString();
                    break;
                case "patientweight":
                    PatientWeight = value.ToString();
                    break;
                //            Case "medicinalproduct" : Drug = value.ToString
                case "manufacturer_name":
                    Manufactures = (List<string>) value;
                    break;
                case "brand_name":
                    BrandNames = (List<string>) value;
                    break;
                case "drugstartdate":
                    DrugStartDate = value.ToString();
                    break;
                case "drugenddate":
                    DrugEndDate = value.ToString();
                    break;
            }
        }

        private void Fetch(Dictionary<string, object> jsonItem)
        {
            foreach (var item in jsonItem)
            {
                var key = item.Key;
                object value = null;

                if (item.Value != null)
                {
                    var checkValue = item.Value.ToString();

                    if (checkValue.StartsWith("{"))
                    {
                        value = MakeDictionaryObject(item.Value.ToString());
                    }
                    else if (checkValue.StartsWith("["))
                    {
                        value = MakeListObject(item.Value.ToString());
                    }
                }

                CheckKeys(key, value);
                AllData.Add(key, value);
            }
        }

        private Dictionary<string, object> MakeDictionaryObject(string data)
        {
            var completeObject = new Dictionary<string, object>();
            var obj = JsonConvert.DeserializeObject<Dictionary<string, object>>(data);

            foreach (var item in obj)
            {
                var key = item.Key;
                object value = null;

                if (item.Value != null)
                {
                    var checkValue = item.Value.ToString();

                    if (checkValue.StartsWith("{"))
                    {
                        value = MakeDictionaryObject(item.Value.ToString());
                    }
                    else if (checkValue.StartsWith("["))
                    {
                        value = MakeListObject(item.Value.ToString());
                    }
                }

                completeObject.Add(key, value);
            }

            return completeObject;
        }

        private List<object> MakeListObject(string data)
        {
            var completeObject = new List<object>();
            var obj = (JArray) (JsonConvert.DeserializeObject(data));

            for (var i = 0; i <= obj.Count - 1; i++)
            {
                var item = obj[i].ToString();
                object value = null;

                if (item.StartsWith("{"))
                {
                    value = MakeDictionaryObject(item.ToString());
                }
                else if (item.StartsWith("["))
                {
                    value = MakeListObject(item.ToString());
                }

                completeObject.Add(value);
            }

            return completeObject;
        }

        #endregion
    }
}

File: SourceCode/Projects/open.fda.gov/ApiDalc/DataObjects/DrugEventItem.vb
﻿<Obsolete>
Public Class DrugEventItem
    Public AllData As New Dictionary(Of String, Object)

    Public Property SafetyReportID As String
    Public Property ReceivedDateString As String
    Public Property SenderOrganization As String


    Public Property PatientSex As String
    Public Property PatientAge As String
    Public Property PatientWeight As String
    Public Property PatientReactions As New List(Of String)

    Public Property DrugStartDate As String
    Public Property DrugEndDate As String

    Public Property Doseage As String
    Public Property Manufactures As New List(Of String)
    Public Property BrandNames As New List(Of String)
    Public Property GenericNames As New List(Of String)

    Public Shared Function Fill(item As Dictionary(Of String, Object)) As DrugEventItem

        Return New DrugEventItem(item)

    End Function

    Public Sub New()
    End Sub


    Private Sub CheckKeys(key As String, value As Object)

        Select Case key

            Case "safetyreportid"
                SafetyReportID = value.ToString
            Case "receivedate"
                ReceivedDateString = value.ToString
            Case "senderorganization"
                SenderOrganization = value.ToString
            Case "patientonsetage"
                PatientAge = value.ToString
            Case "patientweight"
                PatientWeight = value.ToString
                '            Case "medicinalproduct" : Drug = value.ToString
            Case "manufacturer_name"
                Manufactures = CType(value, Global.System.Collections.Generic.List(Of String))
            Case "brand_name"
                BrandNames = CType(value, Global.System.Collections.Generic.List(Of String))
            Case "drugstartdate"
                DrugStartDate = value.ToString
            Case "drugenddate"
                DrugEndDate = value.ToString

        End Select

    End Sub


    Private Sub New(jsonItem As Dictionary(Of String, Object))
        Fetch(jsonItem)
    End Sub

    Private Sub Fetch(jsonItem As Dictionary(Of String, Object))

        For Each item As KeyValuePair(Of String, Object) In jsonItem

            Dim key As String = item.Key
            Dim value As Object = Nothing

            If item.Value IsNot Nothing Then

                Dim checkValue As String = item.Value.ToString

                If checkValue.StartsWith("{") Then

                    value = MakeDictionaryObject(item.Value.ToString)

                ElseIf checkValue.StartsWith("[") Then

                    value = MakeListObject(item.Value.ToString)

                End If

            End If

            CheckKeys(key, value)
            AllData.Add(key, value)

        Next

    End Sub

    Private Function MakeListObject(data As String) As List(Of Object)

        Dim completeObject As New List(Of Object)
        Dim obj As Newtonsoft.Json.Linq.JArray = CType(Newtonsoft.Json.JsonConvert.DeserializeObject(data), Newtonsoft.Json.Linq.JArray)

        For i As Integer = 0 To obj.Count - 1
            Dim item As String = obj(i).ToString
            Dim value As Object = Nothing

            If item.StartsWith("{") Then

                value = MakeDictionaryObject(item.ToString)

            ElseIf item.StartsWith("[") Then

                value = MakeListObject(item.ToString)

            End If

            completeObject.Add(value)

        Next

        Return completeObject

    End Function

    Private Function MakeDictionaryObject(data As String) As Dictionary(Of String, Object)

        Dim completeObject As New Dictionary(Of String, Object)
        Dim obj As Dictionary(Of String, Object) = Newtonsoft.Json.JsonConvert.DeserializeObject(Of Dictionary(Of String, Object))(data)

        For Each item As KeyValuePair(Of String, Object) In obj

            Dim key As String = item.Key
            Dim value As Object = Nothing

            If item.Value IsNot Nothing Then

                Dim checkValue = item.Value.ToString

                If checkValue.StartsWith("{") Then

                    value = MakeDictionaryObject(item.Value.ToString)

                ElseIf checkValue.StartsWith("[") Then

                    value = MakeListObject(item.Value.ToString)

                End If

            End If

            completeObject.Add(key, value)

        Next

        Return completeObject

    End Function


End Class
